### PR TITLE
Dynamic service headers

### PR DESCRIPTION
**Pull request by erwbgy**
First part of #3258

Add support for %CONTOUR_NAMESPACE%, %CONTOUR_SERVICE_NAME% and
%CONTOUR_SERVICE_PORT% dynamic variables and have these expanded like
Envoy dynamic variables are in #3234. (The CONTOUR_ prefix is used so
that there is no possibility of a clash with a future Envoy dynamic
variable.)

If any of these variables are used where they can't be expanded then
they are just passed through literally. For example if
%CONTOUR_SERVICE_NAME% is used at the route level then it is not
expanded and becomes %%CONTOUR_SERVICE_NAME%% when passed to Envoy. This
avoids any Envoy configuration errors due to bad dynamic variables.

Signed-off-by: Keith Burdis <keith.burdis@gs.com>

### CODE REVIEW

File: internal/dag/httpproxy_processor.go [hunk 1]
```diff
@@ -405,6 +406,10 @@ func (p *HTTPProxyProcessor) computeRoutes(
 		delete(p.orphaned, types.NamespacedName{Name: includedProxy.Name, Namespace: includedProxy.Namespace})
 	}
 
+	dynamicHeaders := map[string]string{
+		"CONTOUR_NAMESPACE": proxy.Namespace,
```
Comment (youngnick) [inline internal/dag/httpproxy_processor.go:410]: Seems like this is the only dynamic header added so far? I can't see any sign of the other two, `%CONTOUR_SERVICE_NAME%`, and `%CONTOUR_SERVICE_PORTR%`?

**Review by youngnick — CHANGES_REQUESTED** [review]
The overall shape of this looks great, I think it's just missing:
- support for the extra two fields mentioned in the issue and the description
- testing that the dynamic headers are added correctly. I'd like to see the tests in `builder_test.go` and `policy_test.go` filled out a bit.
- Ideally, we'd have an integration test of some sort that covers testing the user-visible behavior, but that could conceivably be follow-up work.

File: internal/dag/httpproxy_processor.go (hunk 1 above)
Comment (erwbgy) [reply internal/dag/httpproxy_processor.go:410]: They're set on lines 567 and 568 below where the service name and port are available.

**Review by sunjayBhatia — COMMENTED** [review]
> testing that the dynamic headers are added correctly. I'd like to see the tests in builder_test.go and policy_test.go filled out a bit.
 +1

**Comment by skriss** [comment]
Let's make sure we get docs updates for these new variables as well.

File: internal/dag/httpproxy_processor.go (hunk 1 above)
Comment (youngnick) [reply internal/dag/httpproxy_processor.go:410]: Ah, so they are, thanks.

**Comment by youngnick** [comment]
@erwbgy, thanks again for all your work on this. We're due to cut a release before the end of this week, and this is really close. Any chance you could hit the outstanding TODOs?
- expanded testing in `builder_test.go` and `policy_test.go`
- documentation of the available fields in the header replacement HTTPProxy documentation?

**Comment by erwbgy** [comment]
Apologies, I should have added the tests earlier - and they found a bug.  Some documentation changes added as well.

**Comment by erwbgy** [comment]
Add integration test.

**Review by youngnick — APPROVED** [review]
This looks great now, thanks @erwbgy.

**Review by sunjayBhatia — APPROVED** [review]

**Comment by skriss** [comment]
Any objections to getting this one into 1.12?

**Comment by sunjayBhatia** [comment]
> Any objections to getting this one into 1.12?

I would think it is good to go 👍🏽 looks well tested and a small surface area so not super risky

**Comment by erwbgy** [comment]
@skriss I see the [code](https://github.com/projectcontour/contour/blob/release-1.12/internal/dag/httpproxy_processor.go#L409) changes in the 1.12.0 release but it looks like the docs did not get updated?  I see the documentation changes in [main](https://github.com/projectcontour/contour/blob/main/site/docs/main/config/request-rewriting.md) but not [1.12.0](https://github.com/projectcontour/contour/blob/main/site/docs/v1.12.0/config/request-rewriting.md).

**Comment by skriss** [comment]
Ah, shoot, I probably got the 1.12 docs ready before your PR merged and forgot to update them. I'll get a fix in, thanks for pointing this out.

**Comment by skriss** [comment]
See https://github.com/projectcontour/contour/pull/3341.

**Pull request merged**